File: 0010-regular-expression-matching/0010-regular-expression-matching.cs
public class Solution {
    public bool IsMatch(string s, string p) {
        int[,] dp = new int[s.Length+1,p.Length+1];
        return Helper(s,p,s.Length,p.Length,dp);
    }
        private bool Helper(string s, string p, int i, int j, int[,] dp)
    {
        if(i==0 && j==0)
        {
            return true;
        }
        if(j==0)
        {
            return false;
        }
        if(i==0)
        {
            while(j > 0)
            {
                if(p[j-1] == '*')
                {
                    j-=2;
                    continue;
                }
                break;
            }
            return j==0;
        }
        if(dp[i,j] != 0)
        {
            return dp[i,j] == 2;
        }
        if(s[i-1] == p[j-1] || p[j-1] == '.')
        {
            bool temp = Helper(s,p,i-1,j-1,dp);
            dp[i,j] = temp ? 2 : 1;
            return temp;
        }
        else
        {
            if(p[j-1] == '*')
            {
                bool ans = false;
                int ind = i-1;
                if(p[j-2] == s[i-1] || p[j-2] == '.')
                {
                    ans = Helper(s,p,ind,j,dp) || Helper(s,p,i,j-1,dp) || Helper(s,p,i,j-2,dp);
                }
                else
                {
                    ans = Helper(s,p,i,j-1,dp) || Helper(s,p,i,j-2,dp);
                }
                dp[i,j] = ans ? 2 : 1;
                return ans;
            }
            else
            {
                dp[i,j] = 1;
                return false;
            }
        }
    }
}

File: 0010-regular-expression-matching/0010-regular-expression-matching.java
class Solution {
    public boolean isMatch(String s, String p) {
        Boolean[][] dp = new Boolean[s.length()+1][p.length()+1];
        return getAns(s.length(), p.length(), s, p, dp);
    }
    private Boolean getAns(int i, int j, String s, String p, Boolean[][] dp)
    {
        if(i == 0)
        {
            while(j > 0)
            {
                if(p.charAt(j-1) != '*')
                {
                    return false;
                }
                j-=2;
            }
            return true;
        }
        if(j == 0)
        {
            return false;
        }
        if(dp[i][j] != null)
        {
            return dp[i][j];
        }
        if(s.charAt(i-1) == p.charAt(j-1) || p.charAt(j-1) == '.')
        {
            return dp[i][j] = getAns(i-1, j-1, s, p, dp);
        }
        else
        {
            if(p.charAt(j-1) == '*' && (s.charAt(i-1) == p.charAt(j-2) || p.charAt(j-2) == '.'))
            {
                return dp[i][j] = getAns(i-1, j, s, p, dp) || getAns(i, j-1, s, p, dp) || getAns(i, j-2, s, p, dp);
            }
            else if(p.charAt(j-1) == '*')
            {
                return dp[i][j] = getAns(i, j-2, s, p, dp);
            }
        }
        return dp[i][j] = false;
    }
}